### DISCUSSION TITLE
Using Apache Iceberg for table ingestion

### DISCUSSION THREAD

**Discussion by tom-webber** [ministryofjustice/analytical-platform#1615] [Ideas]
## Context
We are building a data ingestion pipeline that handles structured data. It involves a custom solution for the generation, processing, validation, storage, and versioning of data table schemas.

Apache Iceberg is an open source Table format that includes a metadata layer, handling metadata management together with table data. Iceberg includes the following, which covers the functionality we want to provide with our custom metadata handling solution:
- [Schema evolution](https://iceberg.apache.org/docs/latest/evolution/) (changes to the schema over time)
- [Time travel](https://iceberg.apache.org/docs/latest/spark-queries/#time-travel) (ability to query previous versions of the data, accounting for schema evolution)

As well as other schema-handling benefits:
- [Hidden partitioning](https://iceberg.apache.org/docs/latest/partitioning/#icebergs-hidden-partitioning) (users don't need to be aware of partitions to avoid expensive queries)
- [Partition evolution](https://iceberg.apache.org/docs/latest/evolution/#partition-evolution) (changes to the partition method do not require expensive rewrites)

With Schema evolution, we do not need to independently keep track of schema versions as this is inherent to the metadata of each table ([Overview of Iceberg metadata](https://medium.com/snowflake/understanding-iceberg-table-metadata-b1209fbcc7c3)). We can also make use of [snapshot tagging](https://iceberg.apache.org/docs/latest/branching/) if this would be more suitable, though we would likely simply retain all snapshots.

There are a number of other relevant benefits for transitioning to Iceberg Tables across the Data Platform, [particularly for improving query performance through inspection of the partition metadata files at time of query construction](https://iceberg.apache.org/docs/latest/performance/) to pass on only the relevant data files to the query. This will likely be necessary for performant queries of data lake files when using a data visualisation tool such as Superset (otherwise data would need to be loaded into a database to ensure adequate query performance). Previous trials have shown that pairing Superset with Athena query engine on our parquet-only data lake led to long query times that were too long for dashboarding purposes.

## Proposal

We can integrate the inbuilt Iceberg metadata service to our Data Product metadata service to handle processing, validation, storage, and versioning of table metadata. This would hopefully simplify many of our processing stages as they would be inherent to the Iceberg Table metadata.

**Comment by tom-webber** [comment]
[Link for a further overview of Iceberg beyond what is mentioned above, including ACID transactions](https://medium.com/snowflake/how-apache-iceberg-enables-acid-compliance-for-data-lakes-9069ae783b60).

**Comment by murdo-moj** [comment]
One key advantage of iceberg is that it can handle complex loads for dashboarding, where vanilla athena can't. This is the central reason to use iceberg over what we have now, as the data platform will be used to improve tooling and host dashboards.

The schema evolution in iceberg is making us wonder if we should also change how data product versioning works. But **we could leave versioning as it is**, and not use the schema evolution, and still get the performant loads for dashboarding. **This would only be a small change** to current code.

**Comment by LavMatt** [comment]
I'm not sure improved query execution times would be a realised benefit from using iceberg Athena tables as opposed to standard parquet-hive Athena tables.

In the benchmarking exercise that was performed as part of the wider iceberg evaluation from the managed pipelines team, query time was slower for iceberg than equivalent queries using standard parquet hive, especially evident within smaller data ~1GB tables. This sort of size will be much closer to what i expect we'll be dealing with in the MVP, which will be full snapshot loads partitioned by the timestamp relating to the load time. See https://github.com/moj-analytical-services/iceberg-evaluation/blob/main/src/data_derivation/query_performance/benchmark.ipynb for benchmark results.

**Comment by MatMoore** [comment]
I'm very much not an expert on these data file formats, but from reading the docs I get the impression iceberg offers more flexibility than parquet based on the way it stores its own metadata and partition information in addition to the individual parquet files, so it sounds worth exploring further, and running our own benchmarks.

The stuff about [schema/partition evolution](https://iceberg.apache.org/docs/latest/evolution/#evolution) makes me wonder how our current implementation works when there are incremental updates and the schema changes at some point. My assumption is:
- if we just have parquet files we have to reprocess old data to pick up the new schema
- if we use iceberg it would be able to handle this without changing the underlying parquet files

is that correct? And how much of a big deal is this? It sounds like there is an assumption we will be dealing with ~1GB datasets but I'm just curious how confident we are in this, and whether it's likely we will need to support a small number of larger data products, or if we will explicitly discourage people from publishing data products over a certain size.

If we do decide to adopt iceberg, when would we need to/want to implement it?
- If we were to implement it now, do we need to change a lot? Or just the athena queries in the load lambda?
- If we move to dbt, should we change to iceberg before or after that?
- Does it relate to any of our most risky assumptions that we want to test in the alpha?

Also, would this remove our dependence on glue by any chance? If so we might want to reconsider the decision to use that for the schema format we expose in the API.

**Comment by tom-webber** [comment]
Decision:
We are not going to pursue implementation of Iceberg as part of Data as a Product Alpha.

The current design is to ingest tables as full-snapshot every time, and we are not implementing schema evolution. As such the benefits of a Table format such as Iceberg do not outweigh the costs of reimplementation of our existing ingestion pipelines.

There is a plan to move to dbt / create-a-derived-table / table-transformer for the curation process in a later stage of development. This may or may not involve the use of Iceberg ([existing discussions](https://github.com/moj-analytical-services/create-a-derived-table/discussions?discussions_q=iceberg))

If/when we consider implementing schema evolution, we may also revisit this topic. 

[Data format notes](https://dsdmoj.atlassian.net/wiki/spaces/DataPlatform/pages/4536401938/Data+Storage+format)
